File: src/javascript/algorithms/strings/goatLatin/goatLatin.js
// String parsing + per-token transformation  TC: O(N + n^2) SC: O(N)
const VOWELS = new Set(['a','e','i','o','u','A','E','I','O','U']);

function goatLatin(s) {
    return s.split(' ').map((word, i) => {
        if (!VOWELS.has(word[0])) {
            word = word.slice(1) + word[0];
        }
        return word + 'ma' + 'a'.repeat(i + 1);
    }).join(' ');
}

function runTests() {
    const tests = [
        { s: "I speak Goat Latin",        expected: "Imaa peaksmaaa oatGmaaaa atinLmaaaaa" },
        { s: "The quick brown fox jumped", expected: "heTmaa uickqmaaa rownbmaaaa oxfmaaaaa umpedjmaaaaaa" },
        { s: "Each word starts anew",      expected: "Eachmaa ordwmaaa tartssmaaaa anewmaaaaa" },
        { s: "I am",                       expected: "Imaa ammaaa" },
        { s: "mad",                        expected: "admmaa" },
    ];

    tests.forEach(({ s, expected }) => {
        const result = goatLatin(s);
        const status = result === expected ? "PASS" : "FAIL";
        console.log(`[${status}] result=${result} expected=${expected}`);
    });
}

runTests();

File: src/python/algorithms/strings/goatLatin/goat_latin.py
# String parsing + per-token transformation  TC: O(N + n^2) SC: O(N)
VOWELS = set("aeiouAEIOU")

def goat_latin(s):
    words = s.split(" ")
    result = []
    for i, word in enumerate(words):
        if word[0] not in VOWELS:
            word = word[1:] + word[0]
        word += "ma" + "a" * (i + 1)
        result.append(word)
    return " ".join(result)


if __name__ == "__main__":
    test_cases = [
        ("I speak Goat Latin",        "Imaa peaksmaaa oatGmaaaa atinLmaaaaa"),
        ("The quick brown fox jumped", "heTmaa uickqmaaa rownbmaaaa oxfmaaaaa umpedjmaaaaaa"),
        ("Each word starts anew",      "Eachmaa ordwmaaa tartssmaaaa anewmaaaaa"),
        ("I am",                       "Imaa ammaaa"),
        ("mad",                        "admmaa"),
    ]

    for s, expected in test_cases:
        result = goat_latin(s)
        status = "PASS" if result == expected else "FAIL"
        print(f"[{status}] result={result} expected={expected}")
